builder/src/main/java/org/mache/builder/Builder.java

```diff
@@ -0,0 +1,212 @@
+      this.storage = storage;
+      this.storageServers = storageServers;
+      this.cluster = cluster;
+      this.keyspace = keyspace;
+      this.macheType = macheType;
+      this.schemaOption = schemaOption;
+      this.messaging = messaging;
+      this.messagingLocation = messagingLocation;
+      this.topic = topic;
+    }
+
+
+    /**
+     * Returns a mache instance with the correct type info.
+     * Calls out to the service interface to find the StorageProvisioner with the
+     * correct published type and then calls into it to create the mache
+     * then wires in any messaging as necessary
+     * calls through to the messaging service interface to find if that's available
+     *
+     * @return a cache mapping strings to the given type, using the options
+     * specified in the constructor
+     */
+    public ExCache<String, T> macheUp() {
+      final StorageProvisioner storageProvisioner = getStorageProvisionerOrThrow();
+
+      final ExCache<String, T> cache = storageProvisioner.getCache(
+          this.keyspace, macheType, this.schemaOption, this.cluster, this.storageServers);
+
+      if (this.messaging != None) {
+        final MessagingProvisioner messagingProvisioner = getMessagingProvisionerOrThrow();
+        final MQFactory mqFactory = getMqFactoryOrThrowRuntimeException(messagingProvisioner);
+        final MQConfiguration mqConfiguration = () -> this.topic;
+
+        final CacheThingFactory cacheThingFactory = new CacheThingFactory();
+
+        final CacheFactoryImpl cacheFactory = new CacheFactoryImpl(
+            mqFactory, mqConfiguration, cacheThingFactory, new UUIDUtils());
+        return cacheFactory.createCache(cache);
+      }
+      return cache;
+    }
+
+    private MQFactory getMqFactoryOrThrowRuntimeException(MessagingProvisioner messagingProvisioner) {
+      try {
+        return messagingProvisioner.getMQFactory(this.messagingLocation);
+      }
+      catch (IOException | JMSException e) {
+        throw new RuntimeException("Cannot connect to message queue at:[" + this.messagingLocation + "]", e);
+      }
+    }
+
+    private StorageProvisioner getStorageProvisionerOrThrow() {
+      final ServiceLoader<StorageProvisioner> allInClasspath = ServiceLoader.load(StorageProvisioner.class);
+      StorageProvisioner provisioner = null;
+      for (StorageProvisioner storageProvisioner : allInClasspath) {
+        if (this.storage.name().equalsIgnoreCase(storageProvisioner.getStorage())) {
+          provisioner = storageProvisioner;
+        }
+      }
+      if (provisioner == null) {
+        throw new RuntimeException("Cannot find storage provisioner for platform :[" +
+            storage + "].  Please ensure a " + storage + "-ns.jar is present in the classpath");
+      }
+      return provisioner;
+    }
+
+    private MessagingProvisioner getMessagingProvisionerOrThrow() {
+      final ServiceLoader<MessagingProvisioner> allInClasspath = ServiceLoader.load(MessagingProvisioner.class);
```

Comment:
I think this has already been discussed and is on your hit list - but would be best to separate messaging providers into their own modules.